pkg/generators/openapi.go

```diff
@@ -317,14 +339,30 @@ func (g openAPITypeWriter) generate(t *types.Type) error {
 	// Only generate for struct type and ignore the rest
 	switch t.Kind {
 	case types.Struct:
-		if hasOpenAPIDefinitionMethod(t) {
+		hasV2Definition := hasOpenAPIDefinitionMethod(t)
+		hasV2DefinitionTypeAndFormat := hasOpenAPIDefinitionMethods(t)
+		hasV3Definition := hasOpenAPIV3DefinitionMethod(t)
+
+		if hasV2Definition || (hasV3Definition && !hasV2DefinitionTypeAndFormat) {
 			// already invoked directly
 			return nil
 		}
 
 		args := argsFromType(t)
 		g.Do("func "+nameTmpl+"(ref $.ReferenceCallback|raw$) $.OpenAPIDefinition|raw$ {\n", args)
-		if hasOpenAPIDefinitionMethods(t) {
+		switch {
```

Comment:
no separate case for `hasV3Definition`?

Reply:
> no separate case for hasV3Definition?

hasV3Definition case is short-circuited at L329, do you think about adding a case for hasV3Definition and commenting `// this line will never reach` for readability? the branches is getting more complex tho

Reply:
no, we can leave it as it is now.